### PR TITLE
improve prng compile times with loop rolling

### LINKED ISSUE #1172: compiling time grows non-linear w.r.t. the number of random.split ops
A repro code
```
from jax import jit, random
from jax.config import config; config.update("jax_platform_name", "cpu")
import time

def f(key, i):
    for _ in range(i):
        _, key = random.split(key)
    return key

for i in range(17):
    t = time.time()
    key = jit(f, static_argnums=(1,))(random.PRNGKey(0), i).copy()
    print("split {} times takes {}".format(i, time.time() - t))
```
which returns
```
split 0 times takes 0.008060216903686523
split 1 times takes 0.05638694763183594
split 2 times takes 0.13771629333496094
split 3 times takes 0.284271240234375
split 4 times takes 0.4720790386199951
split 5 times takes 0.8609561920166016
split 6 times takes 1.5761287212371826
split 7 times takes 2.615913152694702
split 8 times takes 4.20055079460144
split 9 times takes 6.581374168395996
split 10 times takes 9.83530855178833
split 11 times takes 15.45705771446228
split 12 times takes 20.33790898323059
split 13 times takes 27.418986320495605
split 14 times takes 35.04498267173767
split 15 times takes 44.33582162857056
split 16 times takes 53.186384439468384
```

@mattjj I think that this causes the regression of gamma sampler in our last discussion. The above issue is causing a big problem in compiling models with many random latent variables in NumPyro.

### PR DESCRIPTION
**Pull request by mattjj**
fixes #1172

Code from #1172:

```python
from jax import jit, random
from jax.config import config; config.update("jax_platform_name", "cpu")
import time

def f(key, i):
    for _ in range(i):
        _, key = random.split(key)
    return key

for i in range(17):
    t = time.time()
    key = jit(f, static_argnums=(1,))(random.PRNGKey(0), i).copy()
    print("split {} times takes {}".format(i, time.time() - t))
```

Before (CPU backend):
```
split 0 times takes 0.0584418773651
split 1 times takes 0.0834980010986
split 2 times takes 0.159592151642
split 3 times takes 0.304208993912
split 4 times takes 0.499366998672
split 5 times takes 0.925127029419
split 6 times takes 1.68947792053
split 7 times takes 2.74465990067
split 8 times takes 4.27907800674
split 9 times takes 6.66637420654
split 10 times takes 9.77492880821
split 11 times takes 14.9957740307
split 12 times takes 19.7543628216
split 13 times takes 26.9135191441
split 14 times takes 33.8859760761
split 15 times takes 43.9121148586
split 16 times takes 52.6436569691
```

After 275bf9d, rolling the inner loop (CPU backend):
```
split 0 times takes 0.0326280593872
split 1 times takes 0.174847126007
split 2 times takes 0.271623849869
split 3 times takes 0.443490028381
split 4 times takes 0.658795118332
split 5 times takes 0.810074090958
split 6 times takes 1.0533349514
split 7 times takes 1.34268808365
split 8 times takes 1.60742712021
split 9 times takes 1.83798694611
split 10 times takes 2.12129998207
split 11 times takes 2.35916495323
split 12 times takes 2.70749402046
split 13 times takes 3.08042216301
split 14 times takes 3.3945119381
split 15 times takes 3.67894697189
split 16 times takes 4.16872882843
```

After d857d3f, rolling the outer loop (CPU backend):
```
split 0 times takes 0.0505750179291
split 1 times takes 0.179420948029
split 2 times takes 0.194557905197
split 3 times takes 0.23614692688
split 4 times takes 0.312750101089
split 5 times takes 0.394812107086
split 6 times takes 0.473638772964
split 7 times takes 0.554468870163
split 8 times takes 0.63796877861
split 9 times takes 0.721169948578
split 10 times takes 0.809171199799
split 11 times takes 0.899457931519
split 12 times takes 0.995654821396
split 13 times takes 1.09211206436
split 14 times takes 1.17162895203
split 15 times takes 1.27369189262
split 16 times takes 1.34894180298
```

Not only should compile times be faster in rolling the outer loop, but also it should provide faster execution than rolling the inner loop.

Execution time maybe took a slight hit compared to unrolling everything on master:

```
# on master
In [1]: from jax import random
In [2]: key = random.PRNGKey(0)
jax/lib/xla_bridge.py:109: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
In [3]: timeit random.normal(key, (1000, 1000)).block_until_ready()
The slowest run took 56.75 times longer than the fastest. This could mean that an intermediate result is being cached.
1 loop, best of 3: 5.81 ms per loop
In [4]: timeit random.normal(key, (1000, 1000)).block_until_ready()
100 loops, best of 3: 5.86 ms per loop

# on d857d3f
In [1]: from jax import random
In [2]: key = random.PRNGKey(0)
jax/lib/xla_bridge.py:109: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
In [3]: timeit random.normal(key, (1000, 1000)).block_until_ready()
The slowest run took 44.97 times longer than the fastest. This could mean that an intermediate result is being cached.
1 loop, best of 3: 6.93 ms per loop
In [4]: timeit random.normal(key, (1000, 1000)).block_until_ready()
100 loops, best of 3: 7.45 ms per loop
```

@fehiepsi unless this execution time cost seems onerous (which we can trade off against compile time by doing more unrolling), this seems like a net win to me! Maybe XLA can learn to optimize loops better to close the execution time gap.

If you can make a benchmark for us and the XLA team, we can make it fast!

### CODE REVIEW

File: tests/random_test.py [hunk 1]
```diff
@@ -398,6 +398,7 @@ def testIssue756(self):
       self.assertEqual(onp.result_type(w), onp.float32)
 
   def testNoOpByOpUnderHash(self):
+    raise SkipTest()  # TODO(mattjj): re-enable this test
```
Comment (hawkinsp) [inline tests/random_test.py]: Describe what breaks?

File: jax/random.py [hunk 2]
```diff
@@ -123,38 +123,19 @@ def apply_round(v, rot):
   else:
     x = list(np.split(count.ravel(), 2))
 
-  rotations = onp.uint32([13, 15, 26, 6, 17, 29, 16, 24])
-  ks = [key1, key2, key1 ^ key2 ^ onp.uint32(0x1BD11BDA)]
+  rotations = asarray([[13, 15, 26, 6], [17, 29, 16, 24]], dtype="uint32")
```
Comment (hawkinsp) [inline jax/random.py]: Write the types as `onp.uint32` ?

**Review by hawkinsp — APPROVED** [review]

File: jax/random.py (hunk 2 above)
Comment (mattjj) [reply jax/random.py]: Done.

File: tests/random_test.py (hunk 1 above)
Comment (mattjj) [reply tests/random_test.py]: Fixed! I was being lazy. To make `x[i]` work for a constant `x` I was calling `np.array` to create a DeviceArray, but that was causing op-by-op computation to happen inside `threefry_2x32`. Also that test had cascading effects to break all tests after it. Fixed all!

**Comment by fehiepsi** [comment]
Thanks so much @mattjj ! This works pretty well! The only side effect which I observe is the trick of using `random.split(key, 1)` in gamma sampler does not work anymore with this PR. Anyway, it is minor and I think that it is better to avoid using that trick. I will try to explore warp-based parallelism as proposed in #552 instead.

**Comment by mattjj** [comment]
Can you remind me what that trick does? Maybe we can recover it.

**Pull request merged**